### DISCUSSION TITLE
embed.js is not publicly accessible

### DISCUSSION THREAD

**Discussion by Permanently** [mastodon/mastodon#26877] [Q&A]
I assume this is just a me issue, hence why I ask this.

I'm trying to embed a post, but embed.js just doesn't exist on my instance. No changes from my side at all. The post ends up not showing. Not sure if anyone else has had this issue?

It definitely goes through as well:
`Sep 10 18:10:09 [redacted] bundle[XXXXXX]: [redacted] method=GET path=/embed.js format=js controller=ApplicationController action=raise_not_found status=404 duration=6.49 view=1.31 db=1.03`

When previewing posts, `/packs/js/modals/embed_modal-76dfbae35066e291a273.chunk.js` is loaded. But, embed.js is nowhere to be seen. No idea if they're the same, though.

**Comment by StrangeMinds1960** [comment]
embed.js can be found at /home/mastodon/live/public

**Reply by Permanently** [reply]
Oh yeah, I know that. I can see it in the directory listing. That's not the problem though. The problem is that it's not directly accessible from the domain. I.e., example.com/embed.js returns a 404.

**Reply by StrangeMinds1960** [reply]
Sorry, I assumed from the subject that the file didn't exist.  I've tried the example.com/embed.js test you did and I get the javascript file, so I'm not sure what to suggest

**Comment by renchap** [comment]
This is an issue with your web server (nginx ?) configuration, it should be configured to first try to serve existing files on disk. You can refer to our sample nginx configuration: https://github.com/mastodon/mastodon/blob/main/dist/nginx.conf

If you want Rails to serve your static files (which you should not do in production except if you have a reverse proxy with caching configured), you can set `RAILS_SERVE_STATIC_FILES=1`

**Reply by Permanently** [reply]
I actually use Apache2. I've got [my configuration here](https://gist.github.com/Permanently/f614c2114648135ace72903552c6bf07) (albeit, a bit of a mess).

Got a feeling I might need the env. you mentioned in your second paragraph.

**Reply by renchap** [reply]
You are probably missing a `ProxyPass /embed.js !` directive.

But listing the files like this is probably not a good idea, as it would fail if we add a new static file.

You probably need to do this using rewrite rules, something like that (untested):

```
# If the file does not exists in public/
RewriteCond /home/mastodon/live/public/%{REQUEST_URI} !-f

# Then proxy it to the local Rails server
RewriteRule ^/(.*)$ http://localhost:3000/$1 [P,QSA,L]

# Else, the files will be served from `DocumentRoot` (and we know they exist because of the directives above)
```

The `P` parameter for `mod_rewrite` is a more powerful version of `ProxyPass`, see the doc: https://httpd.apache.org/docs/current/rewrite/flags.html#flag_p

You probably need to adapt the configuration to handle the streaming server as well, but you get the idea. It avoids having to list every existing file / directory from `public` in your configuration, as they may change with Mastodon updates.

**Reply by Permanently** [reply]
I just had a mess around for a little bit on this, and took out the `ProxyPass` bits I had, but seems to not work. Just takes /assets and similar away from being requested.
